operator: Add provision for the user to set proxy env variables in the loki pods

shwetaap opened the pull request:
Signed-off-by: Shweta Padubidri <spadubid@redhat.com>

**What this PR does / why we need it**:
This PR adds a provision for the user to add the proxy environment variables in the loki pods

**Which issue(s) this PR fixes**:

https://issues.redhat.com/browse/LOG-2860

**Special notes for your reviewer**:

**Checklist**
- [ ] Reviewed the `CONTRIBUTING.md` guide
- [ ] Documentation added
- [ ] Tests updated
- [ ] `CHANGELOG.md` updated
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/upgrading/_index.md`

Red-GV reviewed (commented):
I think it looks good. Just a few comments on my end.

periklis commented on operator/apis/loki/v1/lokistack_types.go:
```diff
@@ -323,6 +323,28 @@ type LokiTemplateSpec struct {
+	// +optional
+	// +kubebuilder:validation:optional
+	// +operator-sdk:csv:customresourcedefinitions:type=spec,displayName="Httpproxy"
+	Httpproxy string `json:"httpproxy,omitempty"`
```
```suggestion
	Httpproxy string `json:"httpproxy,omitempty"`
	HTTPProxy string `json:"httpProxy,omitempty"`
```

periklis commented on operator/apis/loki/v1/lokistack_types.go:
```diff
@@ -323,6 +323,28 @@ type LokiTemplateSpec struct {
+	// +optional
+	// +kubebuilder:validation:optional
+	// +operator-sdk:csv:customresourcedefinitions:type=spec,displayName="Httpsproxy"
+	Httpsproxy string `json:"httpsproxy,omitempty"`
```
```suggestion
	Httpsproxy string `json:"httpproxy,omitempty"`
	HTTPSProxy string `json:"httpsProxy,omitempty"`
```

periklis commented on operator/apis/loki/v1/lokistack_types.go:
```diff
@@ -323,6 +323,28 @@ type LokiTemplateSpec struct {
+	// +optional
+	// +kubebuilder:validation:optional
+	// +operator-sdk:csv:customresourcedefinitions:type=spec,displayName="Noproxy"
+	Noproxy string `json:"noproxy,omitempty"`
```
```suggestion
	Noproxy string `json:"noproxy,omitempty"`
	NoProxy string `json:"noProxy,omitempty"`
```

periklis commented on operator/apis/loki/v1/lokistack_types.go:
```diff
@@ -323,6 +323,28 @@ type LokiTemplateSpec struct {
 }
 
+// ClusterProxy is the Proxy configuration when the cluster is behind a Proxy.
+type ClusterProxy struct {
```
I believe this is the right struct to support custom proxy configuration an any kubernetes/proxy environment. For k8s/openshift clusters where the cluster provisioning provides the proxy variables per enviroment, we should add an extra field to pick them up, e.g.:

```go
ReadVarsFromEnv bool `json:"readVarsFromEnv,omitempty"`
```

Ref: https://sdk.operatorframework.io/docs/building-operators/golang/references/proxy-vars/#m-docsbuilding-operatorsgolangreferencesproxy-vars

shwetaap commented:
Unfortunatelly tthis PR closed when I was trying to rebase. I have created another PR addressing the comments - https://github.com/grafana/loki/pull/7384
